01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+## Built-in system actors
+
+The existing built-in system actors will be compiled to WASM and will run entirely in WASM space, thus preventing any associated FFI barrier costs.
+
+For the sake of analogy, built-in system actors in Filecoin receive comparable treatment to "precompiled contracts" in other platforms. They are specified by the protocol, implicitly deployed, addressed at fixed locations/mailboxes, and they evolve through system upgrades.
+
+**Canonical system actors**
+
+This transition to WASM opens up the opportunity to converge on a single codebase of actors for all Filecoin implementations.
+
+In the past, each team would re-implement actors in their language (although some relied on FFI with Go actors, a slower approach). With FVM, a single codebase can be compiled to WASM (its new portable executable form) and be adopted across all implementations.
+
+We acknowledge this strategy has tradeoffs, but we won't elaborate on them here.
+
+## Actor sandbox
+
+The actor sandbox is the tightly constrained and instrumented environment in which Filecoin actor code runs.
+
+The actor sandbox must be formally specified, and potentially versioned. 
+
+User-deployed actors may have to specify the sandbox version they expect, as this facilitates sandbox evolution while preserving backwards compatibility in an obvious manner.
+
+Sandbox responsibilities:
+
+1. To inject/bind/link implicit module imports that provide capabilities such as state access, chain access, syscalls, IPLD codec processing, and more.
+2. To provide a toolbox of common functions and types, such as BigInt arithmetic, parameter unpacking, method dispatch tables, etc.
+3. To manage memory usage and halt execution when it exceeds the limits.
+4. To load input call parameters and handle return values.
+5. To intercept cross contract calls and delegate to the outer environment.
+6. To perform gas accounting by instrumenting bytecode.
+7. To halt execution when available gas is exhausted.
+
+(1) and (2) together form the basis of the **FVM stdlib**: the set of APIs that are universally available for actors to call inside the actor sandbox.
+
+One key constraint is that actor code size/footprint should be kept as low as possible to avoid state bloat and gas costs. (2) is important, as it precludes repetitive/common code from being statically linked to user-deployed actors.
+
+We could also add the ability to deploy libraries as on-chain actors; this would lead to a richer development experience and we'd expect collections of reusable libraries to emerge and be deployed on-chain, for others to consume. This is similar to the use cases of the `CALLDELEGATE` opcode in the EVM. But we'd probably tighten the security characteristics by introducing the notion of "capabilities", such that the calling actor can specify exactly what the target can access.
```

Comment:
There are two different things here:

1. "static" on-chain assets referenced by CID. These wouldn't need any special sandboxing because the smart contract owner would be able to statically validate/determine that they're correct at compile time (the compiler, if deterministic, could even re-use on-chain components if they hash to the expected value).
2. "dynamic" on-chain libraries referenced by address. These are the shared libraries that need to be sandboxed.

Also note: I expect we'll end up using the term "capability" to refer to a slightly different concept. Specifically, I think capabilities will be closely tied with closures. Actor A sends actor B a "closure" that's able to call back into actor A to make some change (at some future point).

Reply:
wow, this is kind of blowing my mind. OOTH, creating yet-another-dependency-management-system seems daunting.

Reply:
I agree.

1. Step 1: Built-in libraries only.
2. Step 2: Arbitrary static libraries.
3. Step 3: Dynamic libraries? Honestly, these aren't _strictly_ necessary as other contracts can be used as libraries with no additional work.

> yet-another-dependency-management-system seems daunting.

Note: The tricky part is dependency resolution which will happen in tooling off-chain. The on-chain part (the part we need to make sure we get "right") will address by CID so it should be easier to ship _something_ and then iterate on the hard parts.